[visionOS] Notify the media player when entering & exiting external playback in LinearMediaPlayer

rsfuller opened the pull request:
#### 7ec7b4de596e9bd0b42dba0732a47ce7e306afd7
<pre>
[visionOS] Notify the media player when entering &amp; exiting external playback in LinearMediaPlayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=291029">https://bugs.webkit.org/show_bug.cgi?id=291029</a>
<a href="https://rdar.apple.com/148287113">rdar://148287113</a>

Reviewed by Jean-Yves Avenard.

External playback mechanism in LinearMediaPlayer used to rely on setting the
videoReceiverEndpoint to NULL to trigger the media player to switch back to
layer based rendering.

Since 42991@main, we no longer get this behavior for free. Instead we need to
notify the media player when we enter/exit external playback similar to
fullscreen/PiP.

One particularly important point for the flow to work is that we must notify
WebGPU that inFullscreenOrPip changed ONLY after the video target changed.

For exiting external playback, this is trivial because LinearMediaPlayer is
setting the videoReceiverEndpoint to NULL directly.

For entering external playback, the video target is not set UNTIL LMK calls
back with setVideoReceiverEndpoint, which happens asynchronously after LMK is
configured for external playback. Thus VideoPresentationInterfaceLMK now waits
for a message that the video receiver endpoint was set before notifying the
media player that fullscreen/PiP state changed.

* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::didEnterExternalPlayback):
(WebCore::HTMLVideoElement::didExitExternalPlayback):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModel::didEnterExternalPlayback):
(WebCore::VideoPresentationModel::didExitExternalPlayback):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::enterExternalPlayback):
(WebKit::VideoPresentationInterfaceLMK::exitExternalPlayback):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationModelContext::didEnterExternalPlayback):
(WebKit::VideoPresentationModelContext::didExitExternalPlayback):
(WebKit::VideoPresentationManagerProxy::didEnterExternalPlayback):
(WebKit::VideoPresentationManagerProxy::didExitExternalPlayback):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.messages.in:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationManager::didEnterExternalPlayback):
(WebKit::VideoPresentationManager::didExitExternalPlayback):

Canonical link: <a href="https://commits.webkit.org/293272@main">https://commits.webkit.org/293272@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a4751851073a81585469b8bf0fea7e11005eb6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103491 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48898 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26451 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74884 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32046 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101373 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88844 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55243 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13646 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6805 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48340 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105864 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25456 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83861 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25829 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85046 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83335 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21055 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27977 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5653 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19108 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25415 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25234 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28550 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26809 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/cocoa/VideoPresentationModel.h:
```diff
@@ -94,6 +94,10 @@ class VideoPresentationModel
     virtual void didExitFullscreen() { };
     virtual void didCleanupFullscreen() { };
     virtual void fullscreenMayReturnToInline() { };
+#if PLATFORM(VISION)
```
```suggestion
#if ENABLE(LINEAR_MEDIA_PLAYER)
```

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
```diff
@@ -114,6 +114,10 @@ class VideoPresentationModelContext final
     void didExitFullscreen() final;
     void didCleanupFullscreen() final;
     void fullscreenMayReturnToInline() final;
+#if PLATFORM(VISION)
```
```suggestion
#if ENABLE(LINEAR_MEDIA_PLAYER)
```

As that’s when those methods are defined

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
```diff
@@ -258,6 +262,10 @@ class VideoPresentationManagerProxy
     void setVideoFullscreenFrame(PlaybackSessionContextIdentifier, WebCore::FloatRect);
     void fullscreenModeChanged(PlaybackSessionContextIdentifier, WebCore::HTMLMediaElementEnums::VideoFullscreenMode);
     void fullscreenMayReturnToInline(PlaybackSessionContextIdentifier);
+#if PLATFORM(VISION)
```
Ditto

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -451,6 +451,22 @@ - (BOOL)prefersStatusBarHidden
         manager->fullscreenMayReturnToInline(m_contextId);
 }
 
+#if PLATFORM(VISION)
```
```suggestion
#if ENABLE(LINEAR_MEDIA_PLAYER)
```

jyavenard commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1530,6 +1546,20 @@ - (BOOL)prefersStatusBarHidden
         page->protectedLegacyMainFrameProcess()->send(Messages::VideoPresentationManager::FullscreenMayReturnToInline(contextId, page->isViewVisible()), page->webPageIDInMainFrameProcess());
 }
 
+#if PLATFORM(VISION)
```
```suggestion
#if ENABLE(LINEAR_MEDIA_PLAYER)
```

jyavenard reviewed (approved):
Should be using 
`#if ENABLE(LINEAR_MEDIA_PLAYER)` in place of PLATFORM(VISION)

LGTM

rsfuller committed:
[visionOS] Notify the media player when entering & exiting external playback in LinearMediaPlayer
https://bugs.webkit.org/show_bug.cgi?id=291029
rdar://148287113

Reviewed by Jean-Yves Avenard.

External playback mechanism in LinearMediaPlayer used to rely on setting the
videoReceiverEndpoint to NULL to trigger the media player to switch back to
layer based rendering.

Since 42991@main, we no longer get this behavior for free. Instead we need to
notify the media player when we enter/exit external playback similar to
fullscreen/PiP.

One particularly important point for the flow to work is that we must notify
WebGPU that inFullscreenOrPip changed ONLY after the video target changed.

For exiting external playback, this is trivial because LinearMediaPlayer is
setting the videoReceiverEndpoint to NULL directly.

For entering external playback, the video target is not set UNTIL LMK calls
back with setVideoReceiverEndpoint, which happens asynchronously after LMK is
configured for external playback. Thus VideoPresentationInterfaceLMK now waits
for a message that the video receiver endpoint was set before notifying the
media player that fullscreen/PiP state changed.

* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::didEnterExternalPlayback):
(WebCore::HTMLVideoElement::didExitExternalPlayback):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModel::didEnterExternalPlayback):
(WebCore::VideoPresentationModel::didExitExternalPlayback):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::enterExternalPlayback):
(WebKit::VideoPresentationInterfaceLMK::exitExternalPlayback):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationModelContext::didEnterExternalPlayback):
(WebKit::VideoPresentationModelContext::didExitExternalPlayback):
(WebKit::VideoPresentationManagerProxy::didEnterExternalPlayback):
(WebKit::VideoPresentationManagerProxy::didExitExternalPlayback):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.messages.in:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationManager::didEnterExternalPlayback):
(WebKit::VideoPresentationManager::didExitExternalPlayback):

Canonical link: https://commits.webkit.org/293272@main
